### PR TITLE
Added support for Nebula Chat model

### PR DESCRIPTION
**Pull request by sharmisthasg**
Description: Added support for Nebula Chat model in addition to Nebula Instruct
Dependencies: N/A
Twitter handle: @Symbldotai

### CODE REVIEW

File: libs/langchain/langchain/chat_models/__init__.py [hunk 1]
```diff
@@ -53,6 +53,7 @@ def __getattr__(name: str) -> None:
     "ChatGooglePalm",
     "ChatMlflow",
     "ChatMLflowAIGateway",
+    "ChatNebula",
```
Comment (eyurtsev) [inline libs/langchain/langchain/chat_models/__init__.py:56]:
```suggestion
```

File: libs/langchain/langchain/chat_models/symbl_nebula.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+from langchain_community.chat_models.symblai_nebula import (
```
Comment (eyurtsev) [inline libs/langchain/langchain/chat_models/symbl_nebula.py:1]: Could you delete this file

File: libs/community/langchain_community/chat_models/symblai_nebula.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    return {
+        "system_prompt": system,
+        "messages" : formatted_messages
+    }
+
+
+class ChatNebula(BaseChatModel):
+    """`Nebula` chat large language models.
+
+    To use, you should have the environment variable ``NEBULA_API_KEY`` set with your API key,
```
Comment (eyurtsev) [inline libs/community/langchain_community/chat_models/symblai_nebula.py:66]: could you include a link to what Nebula is or to its API reference?

File: libs/community/langchain_community/chat_models/symblai_nebula.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+                {
+                    "role": _convert_role(message.type),
+                    "text": message.content,
+                }
+            )
+            """
+            The first message in formatted_messages must be "human"
+            """
+            if formatted_messages[0]["role"] != "human":
+                print("The first message after system prompt must be a human message")
```
Comment (eyurtsev) [inline libs/community/langchain_community/chat_models/symblai_nebula.py:53]: there should be no `print` or `exit()` in the code. If needed raise an exception

**Review by eyurtsev — COMMENTED** [review]
Could you docs to the chat model integrations? https://python.langchain.com/v0.1/docs/integrations/chat/

**Comment by sharmisthasg** [comment]
@eyurtsev I have resolved all your comments barring docs. We do not support tool calling and structured output at this time. Could you please point me to the file where I should note this, for it to appear appropriately in the table?

**Comment by baskaryan** [comment]
> @eyurtsev I have resolved all your comments barring docs. We do not support tool calling and structured output at this time. Could you please point me to the file where I should note this, for it to appear appropriately in the table?

don't need to add nebula to the table but would be nice to have a separate page for it at `docs/docs/integrations/chat/nebula.ipynb`, like this for example: https://python.langchain.com/v0.2/docs/integrations/chat/cohere/

**Comment by sharmisthasg** [comment]
@eyurtsev @baskaryan I have added the documentation. Please let me know if there's anything else required

**Comment by baskaryan** [comment]
> @eyurtsev @baskaryan I have added the documentation. Please let me know if there's anything else required

not passing lint yet. you can run formatting and linting locally with
```bash
cd libs/community
poetry install --with lint,typing
make format
make lint
```

**Comment by sharmisthasg** [comment]
@baskaryan done

**Comment by sharmisthasg** [comment]
Hi @baskaryan is there anything else needed on our end for the final deployment?

**Comment by sharmisthasg** [comment]
Hi @baskaryan @eyurtsev any update on this deployment? The preview looks good, and I think we haven't had any updates in a month. Please let me know if something is needed on my end

**Comment by sharmisthasg** [comment]
@eyurtsev @baskaryan we have an urgent need for this integration, would be very grateful if this merge can be made at the earliest

**Comment by sharmisthasg** [comment]
@eyurtsev @baskaryan could you please tell us what the blocker is, for the merge?

**Comment by ktalamad** [comment]
@eyurtsev @baskaryan @ccurme Folks, could you please let us know if/what blocker there is that needs to be addressed for you to merge this? We have customers blocked on this, and need to get the integration in. Thank you!

**Comment by ccurme** [comment]
> eyurtsev baskaryan ccurme Folks, could you please let us know if/what blocker there is that needs to be addressed for you to merge this? We have customers blocked on this, and need to get the integration in. Thank you!

CI is blocked on the following:

[Docs not matching template](https://github.com/langchain-ai/langchain/actions/runs/9992887014/job/27618982387?pr=21925):
> Document /home/runner/work/langchain/langchain/docs/docs/integrations/chat/symblai_nebula.ipynb does not match the ChatModel Integration page template. Please see https://github.com/langchain-ai/langchain/issues/22296 for instructions on how to correctly format a ChatModel Integration page.

[Mypy / typing errors](https://github.com/langchain-ai/langchain/actions/runs/9992887022/job/27619013464?pr=21925):
> langchain_community/chat_models/symblai_nebula.py:35: error: Incompatible types in assignment (expression has type "Union[str, List[Union[str, Dict[Any, Any]]]]", variable has type "str")  [assignment]
langchain_community/chat_models/symblai_nebula.py:135: error: Incompatible types in assignment (expression has type "str", variable has type "Dict[str, Any]")  [assignment]
langchain_community/chat_models/symblai_nebula.py:172: error: Incompatible types in assignment (expression has type "str", variable has type "Dict[str, Any]")  [assignment]
langchain_community/chat_models/symblai_nebula.py:217: error: Incompatible types in assignment (expression has type "str", variable has type "Dict[str, Any]")  [assignment]
langchain_community/chat_models/symblai_nebula.py:[25](https://github.com/langchain-ai/langchain/actions/runs/9992887022/job/27619013464?pr=21925#step:9:26)2: error: Incompatible types in assignment (expression has type "str", variable has type "Dict[str, Any]")  [assignment]

**Review by ccurme — APPROVED** [review]

**Comment by ccurme** [comment]
@ktalamad @sharmisthasg I got this over but needed to do a small refactor and add ignore a type error. Please take a look and test.

**Comment by sharmisthasg** [comment]
@ccurme thanks for the merge, I was able to test the integration! https://python.langchain.com/v0.2/api_reference/community/chat_models/langchain_community.chat_models.symblai_nebula.ChatNebula.html - this link currently gives a 404, is this something that needs to be added on our end? Is it possible to remove it from the docs while the link is broken?

**Comment by ccurme** [comment]
Hi @sharmisthasg, the link should work now that the API docs have built. The content there is determined by the docstring for `ChatNebula`. You can submit new PRs if you'd like to edit the content at all (feel free to tag me for review).

**Pull request merged**